ARIA IDL updates: enumerated attribute conversions, new "enumerated" wai-aria type, IDL examples

rahimabdi opened the pull request:
Closes #2281 
Closes #2279 

# Mini-Explainer

See [ARIA IDL updates mini-explainer](https://github.com/w3c/aria/wiki/ARIA-IDL-Updates-Mini%E2%80%90Explainer-(ARIA-PR-%232484)).

# Summary of Changes
* Convert following ARIA attributes to enumerated:
  * aria-atomic
  * aria-autocomplete
  * aria-busy
  * aria-checked
  * aria-current
  * aria-disabled
  * aria-expanded
  * aria-haspopup
  * aria-hidden
  * aria-invalid
  * aria-live
  * aria-modal
  * aria-multiline
  * aria-multiselectable
  * aria-orientation
  * aria-pressed
  * aria-readonly
  * aria-required
  * aria-selected
  * aria-sort
* Revise "Mapping WAI-ARIA Value types to languages" to include enumerated attributes
* Revise "6.2.4 Value" value types 
* Update "6.3 ARIA Attributes" section
* Remove obsolete note in "6.3.4 ARIA nullable DOMString Attributes" about ARIA transitioning to non-nullable DOMString?
* Update ARIA IDL usage examples in "6.3.4.1 Example Attribute Usage"
* Add note clarifying "Undefined" state vs. "undefined" string assignment 

# Feature Flag Testing in Safari Tech Preview

You can test the new ARIA IDL reflection behavior with a feature flag in [Safari Tech Preview 228](https://developer.apple.com/documentation/safari-technology-preview-release-notes/stp-release-228). Ensure that "Show features for developers" is enabled (Settings -> Advanced), and then enable the "Enumerated ARIA Attribute Reflection" flag (Settings -> Feature Flags, "JavaScript" section).

The following page can help with quick IDL testing: https://2553411.playcode.io.

# Test, Documentation and Implementation tracking
Once this PR has been reviewed and has consensus from the working group, tests should be written and issues should be opened on browsers. Add N/A and check when not applicable.
* [x] "author MUST" tests: N/A
* [x] "user agent MUST" tests: https://github.com/web-platform-tests/wpt/pull/54119
* [ ] Browser implementations (link to issue or commit):
   * [x] WebKit: https://bugs.webkit.org/show_bug.cgi?id=296851  (Shipping behind feature flag)
   * [ ] Gecko:
   * [ ] Blink:
* [x] Does this need AT implementations? No
* [ ] Related APG Issue/PR:
* [ ] MDN Issue/PR:
* [ ] Further outreach?
   * [x] @rahimabdi outreach to Salesforce contributors (@gregwhitworth et al) re: potential webcompat
   * [ ] Salesforce response pending re: testing the new feature flag in Safari Tech Preview

# Meeting Notes
## ARIA WG & Salesforce Sync (Sept 30 2025)
- Presentation deck
  - [Salesforce/ARIA WG IDL sync (Google Slides)](https://docs.google.com/presentation/d/1exShOV7qZtRFztgP2V9so0kDuhcYCTS1O78e1CkBqfM/edit?usp=sharing)
- Attendees
    - @keithamus, @cookiecrook, @rahimabdi, @gregwhitworth (and other Salesforce team members: Caridy, Eva, Mai, Jason)
- On the LWC side (Lightning Web Components), no anticipated impact unless testing reveals issues
For first-party code, Salesforce framework doesn’t do significant normalization; primary areas of concern would be new behavior around null/undefined/default values
- The Salesforce ARIA IDL polyfill is currently being removed, and it’s only there for backwards compatibility
- There may be potential impact for code packages that aren’t being maintained. For customer code, need a way to communicate the changes (anticipating low impact); Salesforce to investigate an opt-in fallback behavior for those relying on old behavior
- Next steps
    - Salesforce folks to regroup with the ARIA WG in mid-October after completion of impact analysis to LWC and Salesforce first-party code
    - Ideal to wait for Chrome/Firefox to ship the ARIA IDL changes behind a flag as well, and to maintain the flag so it can be toggled off if needed. The goal is to ensure webcompat, so the ARIA WG does not wish to rush the change


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria/pull/2484.html" title="Last updated on Oct 1, 2025, 3:40 PM UTC (5dc7486)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria/2484/38a1bfc...5dc7486.html" title="Last updated on Oct 1, 2025, 3:40 PM UTC (5dc7486)">Diff</a>

rahimabdi commented on index.html:
```diff
@@ -12308,21 +12294,25 @@ <h2>Definitions of States and Properties (all aria-* attributes)</h2>
               </tr>
             </tbody>
           </table>
+          <p>The attribute's <a data-cite="html/common-microsyntaxes.html#missing-value-default">missing value default</a> and <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a> are both the Missing state.</p>
```
Migrated comment from @keithamus ([original](https://github.com/w3c/aria/pull/2413/files#r1983915442)):

> HTML has a concept of states that don't map to a keyword, but each state must have a definition, the definition allows mapping back to a keyword, or null. As an example the `popover` attribute has a missing value default of [no-popover state](https://html.spec.whatwg.org/multipage/popover.html#attr-popover-none-state) which let's the spec say things like "if the popover attribute is in the no-popover state".
> 
> Additionally, traditional enumerated attributes will map an invalid state back to a particular keyword; for example `popover`'s invalid attribute state is the `manual` state, which maps to the keyword `"manual"`. It's important to know what the invalid state maps back to because it determines what value is returned in the IDL for those values.
> 
> Note: not all missing values return null (e.g. the `kind` attribute on `<track>` elements has a missing value default of the subtitles state, so a track's `.kind` property returns `"subtitles"` even if the attribute is not present, another example is the `type` attribute on `<input>` which defaults to `"text"`).
> 
> 
> With that said it's possible to be in "no state". In fact, an enumerated attribute without a missing value default will be in "no state", as will invalid. No state, AIUI, will always map to null.
> 
> So here, the simplest thing for us to do here, rather than the implied missing state is to remove this, and lean on the _lack of_ a missing/invalid state returning the no state. IOW this line could simply be removed:
> 
> ```suggestion
> ```
> 
> If you think missing the definitions create more ambiguity, then we can instead do something like:
> 
> ```suggestion
> 			<p>The attribute has no <a data-cite="html/common-microsyntaxes.html#missing-value-default">missing value default</a> or <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a>.</p>
> ```

rahimabdi replied:
@keithamus My understanding is that, in order to keep backwards compatibility, we'll need to introduce a state without a keyword that defines the null/undefined value. I don't fully understand the difference between the lack of a missing/invalid value default state (such as "Missing") and explicitly defining it as a state (e.g., like popover's "no popover state" which is null) as they would both be null. Happy to change it for spec clarity but would appreciate you clarifying!

keithamus replied:
Can you help me understand what backwards-compatibility you're concerned about? AFAICT setting the IDL reflection to `null`/`undefined` removes the attribute, e.g. if I run `.ariaAtomic = null` then `.hasAttribute('aria-atomic')` is false. This is in line with other enumerated attributes so I can't tell why we'd need this additional state.

annevk replied:
Given the definition at https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#keywords-and-enumerated-attributes I think Keith is correct that there's no difference between an explicit "missing state" and HTML's "no state".

However, I think that for `DOMString?` IDL attributes we would end up hitting an assert in https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#limited-to-only-known-values without an explicit state.

So we would either have to change that wording or go with an explicit state. I personally rather like having an explicit state for all conditions of the attribute (and it seems to me code would have to deal with that regardless of whether it's "missing state" or "no state").

rahimabdi replied:
Thanks for chiming in @annevk.

@keithamus I don't feel strongly either way but would like to hear your thoughts.

keithamus replied:
I don't feel strongly on this. That @annevk mentions `DOMString?` hits an assert is valid, so I think keeping the explicit `Missing` state seems reasonable here. Perhaps we can add a note calling out the reason that the `Missing` state exists? Something to the effect of it providing an explicit alternative to the `no state`, but that implementers may elide the state as it's functionally identical to a `no state`.

rahimabdi replied:
@keithamus I've added this note in the new "Enumerated Attribute Value" section and would welcome your thoughts:

> Some ARIA attributes support an Undefined state, which is equivalent to having no state (i.e., null). User agent implementations may omit the Undefined state since the lack of an explicit state results in the same behavior. For more information about determining the state of enumerated attributes, see Keyword and enumerated attributes (which links to ("html/common-microsyntaxes.html#keywords-and-enumerated-attributes").

keithamus replied:
I think this is useful to call out, thanks for this. I'll mark this piece resolved.

rahimabdi commented on index.html:
```diff
@@ -12771,30 +12773,36 @@ <h2>Definitions of States and Properties (all aria-* attributes)</h2>
               </tr>
             </tbody>
           </table>
+          <p>The attribute's <a data-cite="html/common-microsyntaxes.html#missing-value-default">missing value default</a> and <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a> are both the Undefined state.</p>
```
Migrated comment from @keithamus ([original](https://github.com/w3c/aria/pull/2413/files#r1983923427)):

I don't think we want the missing value default here to return the Undefined state, that will mean the `.ariaChecked` property will return `"undefined"` when `aria-checked` isn't on the element. I presume we want to return `null` instead? Browsers currently return null, so missing should be "no state", i.e. removed:

```suggestion
			<p>The attribute has no <a data-cite="html/common-microsyntaxes.html#missing-value-default">missing value default</a> and its <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a> are both the Undefined state.</p>
```

rahimabdi replied:
@keithamus Updated "Undefined state" to reference that it is equivalent to null.

rahimabdi commented on index.html:
```diff
@@ -13963,41 +13991,50 @@ <h2>Definitions of States and Properties (all aria-* attributes)</h2>
               </tr>
             </tbody>
           </table>
+          <p>The attribute's <a data-cite="html/common-microsyntaxes.html#missing-value-default">missing value default</a> is the Undefined state, and its <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a> is the Menu state.</p>
```
Migrated comment from @keithamus ([original](https://github.com/w3c/aria/pull/2413/files#r1983930085)):

Again, here, the missing value default being the `Undefined` state means `.ariaHasPopup` will return `"undefined"` which I don't think is desirable. Instead:

```suggestion
			<p>The attribute has no <a data-cite="html/common-microsyntaxes.html#missing-value-default">missing value default</a>, and its <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a> is the Menu state.</p>
```

rahimabdi replied:
@keithamus Thanks! `aria-haspopup` doesn't actually support an undefined state so I've removed this. The updated enumerated attribute for aria-haspopup:

Keyword → State Mapping:
- false, (empty string) → False
- true, menu → Menu
- listbox → Listbox
- tree → Tree
- grid → Grid
- dialog → Dialog

Missing Value Default:
- (null)

Invalid Value Default:
- false

rahimabdi commented:
@keithamus:

> @rahimabdi I've left a couple of review notes which I think apply in a lot of places, but rather than inundating you with review comments hopefully you can take the couple here and pattern match across the rest of the spec.

Thanks! I'll review those comments (which are migrated to this PR now) and respond.

rahimabdi commented:
Re-added reviewers @pkra @spectranaut @cookiecrook @keithamus @scottaohara.

rahimabdi commented:
CC @pkra @spectranaut @cookiecrook @keithamus @scottaohara @annevk 

This PR is ready for review! I've made a number of updates, mainly:
- Cleaned up enumerated attribute keywords/states per @keithamus's feedback.
- Changed `aria-atomic`'s Missing state to Undefined, to align with other ARIA attributes that have an "Undefined" state
- Added more IDL usage examples
- Added a note clarifying that Undefined is equivalent to "no state"

If helpful, I also created this tool to simulate how the new ARIA IDL works for enumerated attributes: https://rahimabdi.github.io/aria-idl-simulator/.

annevk commented on index.html:
```diff
@@ -11932,52 +11924,76 @@ <h3>Value</h3>
           <p>
-            When the ARIA attribute definition includes a table listing the attribute's allowed <span>values</span>, that attribute is a multi-value nullable attribute. Each value in the table is a
-            keyword for the attribute, mapping to a state of the same name.
+            Some ARIA attributes support an Undefined state, which is equivalent to having no state (i.e., null). User agent implementations may omit the Undefined state since the lack of an explicit
+            state results in the same behavior. For more information about determining the state of enumerated attributes, see
```
I think this sentence adds more confusion than it resolves. Especially since you cannot just leave it out at the moment without hitting an assert.

rahimabdi replied:
@annevk @cookiecrook Got it, I've removed this sentence: "User agent implementations may omit the Undefined state since the lack of an explicit state results in the same behavior."

@keithamus Hopefully the link to the HTML spec (which defines the state computation algorithm) is sufficient. 

> Do we ever anticipate a time where the explicit "undefined" content attribute values would need to map to an Undefined state, separately than the Null state? For example, I think the only scenarios where an undefined is allowed are with the undefined/true/false/[/mixed] enumerated values, but since the Missing Default and Invalid Default are the same one those, there is no impact... Can we foresee any scenario where those would be different? For example, if undefined was an allowed token value of aria-invalid? (Presumably we just need to always avoid that, which seems okay.)

As @keithamus also mentioned, the "undefined" content attribute value will map to the invalid value default (effectively treated as invalid), and fortunately there's no need for any ARIA attribute to have an explicit "undefined "keyword (nor an Undefined state mapping to "undefined" the keyword). I plan to add a note on this in this PR.

Here are all the attributes which currently support a value of "undefined" per spec, all mapping to their respective invalid value defaults and equivalent with behavior currently defined in spec:
- aria-checked ("undefined" maps to null)
- aria-expanded ("undefined" maps to null)
- aria-hidden ("undefined" maps to False state) 
- aria-orientation ("undefined" maps to null)
- aria-pressed ("undefined" maps to null)
- aria-selected ("undefined" maps to null)

Note: Each of the 6 attributes also support the empty string as a keyword, so `""` and "undefined" are the same. 

Only two enumerated ARIA attributes have differing invalid/missing value defaults and they don't support "undefined" however, any invalid value provided to them should result in a value of "true" per spec (now that they're enumerated, this is handled by the invalid value default):
- aria-current
    - invalid value default = True state
    - missing value default = False state
- aria-invalid
    - invalid value default = True state
    - missing value default = False state

annevk commented on index.html:
```diff
@@ -11932,52 +11924,76 @@ <h3>Value</h3>
-            Default values from the ARIA values tables MUST NOT reflect to IDL as the <a data-cite="html/common-microsyntaxes.html#missing-value-default">missing value default</a> or the
-            <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a> for the attribute. On getting, a missing ARIA attribute will return <code>null</code>. ARIA
-            attributes are not validated on get. If an ARIA value is invalid, on getting, it will return its set value as a literal string, and will not return an invalid value default.
+            For non-enumerated attributes, default values from the ARIA values tables MUST NOT reflect to IDL as the
```
This seems okay to state as a note, but it shouldn't be a requirement. HTML already requires this.

rahimabdi replied:
Thanks! Updated this to be an informative note (added `class="note"` like other spec examples).

annevk reviewed (commented):
One thing that's nice in HTML is that all the states are a `<dfn>` and therefore also can be linked. But maybe that's best addressed in a follow-up.

pkra commented on index.html:
```diff
@@ -12761,7 +12802,7 @@ <h2>Definitions of States and Properties (all aria-* attributes)</h2>
               <tr>
                 <th class="state-value-head" scope="row">Value:</th>
-                <td class="state-value"><a href="#valuetype_tristate">tristate</a></td>
+                <a href="#valuetype_enumerated">enumerated</a>
```
This seems better
```suggestion
                <td class="state-value"><a href="#valuetype_enumerated">enumerated</a></td>
```

rahimabdi replied:
Fixed, thank you.

cyns commented:
@aleventhal you may want to review this

pkra commented on index.html:
```diff
@@ -12412,32 +12435,42 @@ <h2>Definitions of States and Properties (all aria-* attributes)</h2>
               </tr>
               <tr>
-                <th class="value-name" scope="row"><strong class="default">none (default)</strong></th>
+                <th class="value-name" scope="row">none</th>
```
Is it intentional that `(default)` is dropped here? (And in 18 other instances.)

rahimabdi replied:
I've reverted all of these to what is reflecting in main (showing default in parenthesis).

Nice idea @cookiecrook! @pkra perhaps an agenda topic for the Editor's meeting?

spectranaut commented:
Discussed in today's meeting: https://www.w3.org/2025/05/08-aria-minutes.html#7d9a

cyns reviewed (approved):
lgtm! this is much more clear, thank you

keithamus commented on index.html:
```diff
@@ -12308,21 +12323,29 @@ <h2>Definitions of States and Properties (all aria-* attributes)</h2>
+          <p>
+            For <a data-cite="html/common-dom-interfaces.html#reflect">reflection</a> purposes, the attribute's
+            <a data-cite="html/common-microsyntaxes.html#missing-value-default">missing value default</a> is the Undefined state (null) that has no associated keyword, and its
+            <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a> is the False state.
```
As the `False` state has two keywords we will need to specify the canonical keyword for this state;

```suggestion
            <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a> is the False state. For the purposes of reflection, the canonical keyword for the False state is the `false` keyword.
```

(You might want to link to "reflection", "canonical keyword", etc).

This is the same for other properties like `aria-busy`, `aria-checked` and so on - I won't leave a note for each.

keithamus replied:
https://github.com/whatwg/html/pull/11303 has landed so we can consider this resolved, I believe.

keithamus commented on index.html:
```diff
@@ -12308,21 +12323,29 @@ <h2>Definitions of States and Properties (all aria-* attributes)</h2>
+          <p>
+            For <a data-cite="html/common-dom-interfaces.html#reflect">reflection</a> purposes, the attribute's
+            <a data-cite="html/common-microsyntaxes.html#missing-value-default">missing value default</a> is the Undefined state (null) that has no associated keyword, and its
+            <a data-cite="html/common-microsyntaxes.html#invalid-value-default">invalid value default</a> is the False state.
```
So I believe this kind of change to the spec is web observable, which means it will need changes to the implementations and we need to ensure compatibility for web developers who may depend on the existing behaviour. Unless I am missing that somewhere we clarify that reflected enumerated properties should only return the value of the content attribute?

In this example the web observable piece is that the "invalid value default" is the `False` state, which means that any invalid value would map to the `False` (for which I presume the keyword is `"false"` - see above).

This means pasting this HTML into your URL bar:

```
data:text/html,<body id=b><div id=d aria-atomic=foo><script>b.append(d.ariaAtomic)</script>
```

Should render `false` on the page. But right now it renders `foo`.

I think this will be the same for all of these properties, e.g. `aria-autocomplete`, `aria-checked` and so on.

annevk replied:
Right, we should have tests for this and such, but this is very much the motivation for making this change. This makes these IDL attributes more meaningful and enables them to be used for feature testing going forward.

cookiecrook replied:
Seems like we should add the tentative WPT tests, and implement in engines behind a runtime flag to review potential impact on the Salesforce polyfill (Keith recalled this example) and any other places this might affect webcompat. @nolanlawson FYI ^

gregwhitworth replied:
@nolanlawson @divmain @wjhsf @cookiecrook Following up here at SF to take a look into this change. Much appreciated.

keithamus commented on index.html:
```diff
@@ -12771,30 +12812,40 @@ <h2>Definitions of States and Properties (all aria-* attributes)</h2>
               </tr>
               <tr>
-                <th class="value-name" scope="row"><strong class="default">undefined</strong> (default)</th>
+                <th class="value-name" scope="row">undefined, empty string ("")</th>
```
We could reasonably drop the `undefined` keyword here:

```suggestion
                <th class="value-name" scope="row">empty string ("")</th>
```

If someone set `"undefined"` then it would map to the invalid value default, which is the undefined state. The only reason _not_ to do this is if you want the canonicalization steps to explicitly return `undefined` rather than the empty string.

The same is true for other properties, e.g. `aria-current`, and so on.

spectranaut commented:
Discussed briefly: https://www.w3.org/2025/06/26-aria-minutes.html#768c

Next step to write tentative tests?

annevk commented:
I think supporting the empty string as a valid keyword makes sense, but "`undefined`" not so much.

Commit:
Update 'indicates' definition

Commit:
Update note on boolean-like attrs

rahimabdi commented on index.html:
```diff
@@ -623,9 +623,8 @@ <h1>Important Terms</h1>
-              <a href="#valuetype_true-false">true/false</a>, <a href="#valuetype_true-false-undefined">true/false/undefined</a>, <a href="#valuetype_tristate">tristate (true/false/mixed)</a>, a
-              single named <a href="#valuetype_token">token</a>, or a <a href="#valuetype_token_list">token list</a>.
+              Used in an attribute description to denote that the value <a href="#propcharacteristic_value">type</a> is a named token or otherwise token-like, including a single named
+              <a href="#valuetype_token">token</a>, a <a href="#valuetype_token_list">token list</a>, or <a href="#valuetype_enumerated">enumerated</a>.
```
Agreed, I've added a note clarifying this.

Commit:
More updates

Commit:
Update examples

Commit:
Minor fixes

Commit:
Add enumerated attr definition

Commit:
Additional minor fixes

Commit:
Add canonical keywords

Commit:
Add not on undefined value versus Undefined state

rahimabdi commented:
I've made many updates to this PR addressing outstanding feedback. I've also updated the PR description with a link to an unofficial explainer, details on how to test (via Safari feature flag) and more.

I will Agenda+ to discuss these outstanding questions:
- Mapping of `aria-haspopup`: should "true" and "menu" be distinct states in terms of IDL?
- Specifying undefined in the attribute Values table: does it look OK?
- Feedback on the new "Relationship Between the Undefined State and "undefined" String Value" section
- Should all ARIA enumerated attributes that support null be specified as an Undefined state? This question applies only to `aria-atomic` and `aria-haspopup`.

rahimabdi commented:
@keithamus @cookiecrook and I met with the Salesforce team on Sept 30 2025 to discuss impact/risks for Salesforce frameworks. I've captured meeting notes/next steps in the "Meeting Notes" section in the PR description.

pkra commented:
Discussed in the last editors' call https://www.w3.org/2026/05/13-aria-editors-minutes.html#a346
